cdc/log.cc

```diff
@@ -722,15 +914,46 @@ class transformer final {
                         res.set_cell(log_ck, *dst, atomic_cell::make_live(*dst->type, ts, *value, _cdc_ttl_opt));
                     }
 
-                    if (has_pirow) {
-                        value = get_preimage_col_value(cdef, pirow);
+                    bytes_opt prev;
 
+                    if (has_pirow) {
+                        prev = get_preimage_col_value(cdef, pirow);
                         assert(std::addressof(res.partition().clustered_row(*_log_schema, *pikey)) != std::addressof(res.partition().clustered_row(*_log_schema, log_ck)));
                         assert(pikey->explode() != log_ck.explode());
-                        res.set_cell(*pikey, *dst, atomic_cell::make_live(*dst->type, ts, *value, _cdc_ttl_opt));
+                        res.set_cell(*pikey, *dst, atomic_cell::make_live(*dst->type, ts, *prev, _cdc_ttl_opt));
+                    }
+
+                    if (postimage) {
+                        // keep track of actually assigning this already
+                        columns_assigned.emplace(id);
+                        // don't merge with pre-image iff column delete
+                        if (is_column_delete) {
+                            prev = std::nullopt;
+                        }
+                        if (cdef.is_atomic() && !is_column_delete && value) {
+                            res.set_cell(*poikey, *dst, atomic_cell::make_live(*dst->type, ts, *value, _cdc_ttl_opt));
+                        } else if (!cdef.is_atomic() && (value || (deleted_elements && prev))) {
+                            auto v = visit(*cdef.type, make_visitor([&] (const auto& type) -> bytes {
```

Comment:
`make_visitor` is not needed here, the generic lambda is a valid visitor here by itself.

Reply:
Quite.